File: scripts/sync_attendance.py
"""
OPTIONAL fallback sync script (Python).
The app now syncs automatically via the built-in API route — you don't need this script.
Use it only if the app server is not running and you need a one-off manual sync.

Requirements:
  pip install pyzk supabase python-dotenv

Usage:
  python scripts/sync_attendance.py           # full sync
  python scripts/sync_attendance.py --show    # print raw device data only
"""

import sys
from pathlib import Path
from zk import ZK
from supabase import create_client
from datetime import timezone, timedelta
from dotenv import load_dotenv
import os

# Read credentials from Next.js .env.local (same vars the app uses)
load_dotenv(Path(__file__).parent.parent / ".env.local")
SUPABASE_URL = os.getenv("NEXT_PUBLIC_SUPABASE_URL", "")
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY", "")
ZK_IP        = os.getenv("ZK_DEVICE_IP", "192.168.1.101")
ZK_PORT      = int(os.getenv("ZK_DEVICE_PORT", "4370"))

IST = timezone(timedelta(hours=5, minutes=30))


def main():
    if not SUPABASE_URL or not SUPABASE_KEY:
        print("ERROR: NEXT_PUBLIC_SUPABASE_URL / SUPABASE_SERVICE_ROLE_KEY not found in .env.local")
        sys.exit(1)

    show_only = "--show" in sys.argv
    print(f"Connecting to {ZK_IP}:{ZK_PORT} ...")
    conn = ZK(ZK_IP, port=ZK_PORT, timeout=10).connect()

    try:
        if show_only:
            for u in conn.get_users():
                print(f"  uid={u.uid}  id={u.user_id!r:>6}  name={u.name!r:30}  card={u.card}")
            att = conn.get_attendance()
            for a in list(att)[-10:]:
                print(f"  id={a.user_id!r:>6}  time={a.timestamp}  status={a.status}")
            print(f"Total attendance records: {len(att)}")
            return

        sb = create_client(SUPABASE_URL, SUPABASE_KEY)

        users = conn.get_users()
        if users:
            rows = [{"bio_user_id": str(u.user_id), "name": u.name.strip() or f"User {u.user_id}",
                     "device_uid": u.uid, "privilege": u.privilege, "card_no": u.card or 0} for u in users]
            sb.table("staff").upsert(rows, on_conflict="bio_user_id").execute()
            print(f"Staff synced: {len(rows)}")

        att = conn.get_attendance()
        if att:
            records = []
            for a in att:
                t = a.timestamp.replace(tzinfo=IST) if a.timestamp.tzinfo is None else a.timestamp
                records.append({"bio_user_id": str(a.user_id), "punch_time": t.isoformat()})
            for i in range(0, len(records), 500):
                sb.table("attendance_logs").upsert(records[i:i+500], on_conflict="bio_user_id,punch_time").execute()
            print(f"Attendance synced: {len(records)}")
    finally:
        conn.disconnect()


if __name__ == "__main__":
    main()

File: scripts/sync-attendance.js
/**
 * Local sync script — run this on the shop PC to pull attendance from the
 * biometric device and push it to Supabase.
 *
 *   node scripts/sync-attendance.js
 *
 * Reads credentials from .env.local (same file the Next.js app uses).
 * No extra install needed — node-zklib is already in package.json.
 */

const fs   = require("fs");
const path = require("path");

// ── Load .env.local ─────────────────────────────────────────────────────────
const envPath = path.join(__dirname, "..", ".env.local");
if (fs.existsSync(envPath)) {
  fs.readFileSync(envPath, "utf8")
    .split("\n")
    .forEach((line) => {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) return;
      const eq = trimmed.indexOf("=");
      if (eq === -1) return;
      const key = trimmed.slice(0, eq).trim();
      const val = trimmed.slice(eq + 1).trim().replace(/^["']|["']$/g, "");
      if (!process.env[key]) process.env[key] = val;
    });
} else {
  console.error("ERROR: .env.local not found at", envPath);
  process.exit(1);
}

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;
const ZK_IP        = process.env.ZK_DEVICE_IP   || "192.168.1.101";
const ZK_PORT      = parseInt(process.env.ZK_DEVICE_PORT || "4370");

if (!SUPABASE_URL || !SUPABASE_KEY) {
  console.error("ERROR: NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY missing in .env.local");
  process.exit(1);
}

// ── Supabase REST helpers ────────────────────────────────────────────────────
async function supabaseUpsert(table, rows, onConflict, prefer = "resolution=merge-duplicates") {
  const res = await fetch(
    `${SUPABASE_URL}/rest/v1/${table}?on_conflict=${onConflict}`,
    {
      method: "POST",
      headers: {
        apikey: SUPABASE_KEY,
        Authorization: `Bearer ${SUPABASE_KEY}`,
        "Content-Type": "application/json",
        Prefer: prefer,
      },
      body: JSON.stringify(rows),
    }
  );
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Supabase error on ${table}: ${text}`);
  }
}

// ── Main ────────────────────────────────────────────────────────────────────
async function main() {
  const ZKLib = require("node-zklib");
  const zk = new ZKLib(ZK_IP, ZK_PORT, 10000, 4000);

  console.log(`Connecting to ZK device at ${ZK_IP}:${ZK_PORT} ...`);
  await zk.createSocket();
  console.log("Connected.");

  try {
    // Sync staff
    const { data: users = [] } = await zk.getUsers();
    if (users.length) {
      // Step 1: insert NEW staff only (ignore existing — preserves manually-edited names)
      const newRows = users.map((u) => ({
        bio_user_id: String(u.userId),
        name:        u.name?.trim() || `User ${u.userId}`,
        device_uid:  u.uid   ?? null,
        privilege:   u.role  ?? 0,
        card_no:     u.cardno ?? 0,
      }));
      await supabaseUpsert("staff", newRows, "bio_user_id", "resolution=ignore-duplicates");

      // Step 2: update device fields only for ALL staff (never touches name)
      const deviceRows = users.map((u) => ({
        bio_user_id: String(u.userId),
        device_uid:  u.uid   ?? null,
        privilege:   u.role  ?? 0,
        card_no:     u.cardno ?? 0,
      }));
      await supabaseUpsert("staff", deviceRows, "bio_user_id", "resolution=merge-duplicates");

      console.log(`Staff synced: ${newRows.length}`);
      newRows.forEach((r) => console.log(`  [${r.bio_user_id}] ${r.name}`));
    }

    // Sync attendance
    const { data: logs = [] } = await zk.getAttendances();
    const records = logs
      .filter((a) => a.deviceUserId)
      .map((a) => ({
        bio_user_id:  String(a.deviceUserId),
        punch_time:   (a.recordTime instanceof Date ? a.recordTime : new Date(a.recordTime)).toISOString(),
        punch_status: a.state ?? null,
        punch_type:   a.type  ?? null,
      }));

    for (let i = 0; i < records.length; i += 500) {
      await supabaseUpsert("attendance_logs", records.slice(i, i + 500), "bio_user_id,punch_time");
    }
    console.log(`Attendance records synced: ${records.length}`);
    console.log("Done.");
  } finally {
    await zk.disconnect();
  }
}

main().catch((err) => {
  console.error("SYNC FAILED:", err.message || err);
  process.exit(1);
});
